test/unit/mt_membar.c

```diff
@@ -0,0 +1,189 @@
+
+        if (tid == 1) {
+            shmem_int_wait_until(&shared_dest_2, SHMEM_CMP_EQ, one);
+            shmem_int_atomic_add(&result, shared_dest_1 + shared_dest_2, me);
+            shared_dest_2 = 0;
+            shared_dest_1 = 0;
+            shmem_fence();
+        }
+    }
+
+    pthread_barrier_wait(&fencebar);
+    if (0 == tid) shmem_barrier_all();
+    pthread_barrier_wait(&fencebar);
+
+    if (tid == 0) {
+        errors += result == T * ITER ? 0 : 1;
+        if (result != T * ITER) {
+            printf("ERROR in FENCE test from %d : result = %d, expected = %d\n",
+                    me, result, T * ITER);
+        }
+        result = 0;
+	shared_dest_1 = 0;
+	shared_dest_2 = 0;
+    }
+
+    pthread_barrier_wait(&fencebar);
+    if (0 == tid) shmem_barrier_all();
+    pthread_barrier_wait(&fencebar);
+
+    /* TEST LOCK */
+    for (i = 0; i < ITER; i++) {
+        if (tid == 0) {
+            shmem_set_lock(&lock);
+            shared_dest_1 = 1;
+            shmem_clear_lock(&lock);
+            shared_dest_2 = 1;
+            shmem_int_wait_until(&shared_dest_2, SHMEM_CMP_EQ, zero);
+        }
+
+        if (tid == 1) {
+            shmem_int_wait_until(&shared_dest_2, SHMEM_CMP_EQ, one);
+            shmem_int_atomic_add(&result, shared_dest_1, me);
+            shared_dest_1 = 0;
+            shared_dest_2 = 0;
+        }
+    }
+
+    pthread_barrier_wait(&fencebar);
+    if (0 == tid) shmem_barrier_all();
+    pthread_barrier_wait(&fencebar);
+
+    if (tid == 0) {
+        errors += result == ITER ? 0 : 1;
+        if (result != ITER) {
+            printf("ERROR in LOCK test from %d : result = %d, expected = %d\n",
+                    me, result, T * ITER);
+        }
+        result = 0;
+    }
+
+    pthread_barrier_wait(&fencebar);
+    if (0 == tid) shmem_barrier_all();
+
+    return NULL;
+}
+
+int main(int argc, char **argv) {
+    int tl, i, ret;
+    pthread_t threads[T];
+    int t_arg[T];
+
+    ret = shmem_init_thread(SHMEM_THREAD_MULTIPLE, &tl);
+
+    if (tl != SHMEM_THREAD_MULTIPLE || ret != 0) {
+        printf("Init failed (requested thread level %d, got %d, ret %d)\n", 
+               SHMEM_THREAD_MULTIPLE, tl, ret);
+        if (ret == 0) {
+            shmem_global_exit(1);
+        } else {
+            return ret;
+        }
+    }
+
+    me = shmem_my_pe();
+    npes = shmem_n_pes();
+
+    for (i = 0; i < _SHMEM_REDUCE_SYNC_SIZE; i++) {
+        pSync[i] = _SHMEM_SYNC_VALUE;
+    }
+
+    pthread_barrier_init(&fencebar, NULL, T);
+
+    if (me == 0) {
+        printf("Starting multi-threaded test on %d PEs, %d threads/PE\n", npes, T);
+    }
+
+    for (i = 0; i < T; i++) {
+        int err;
+        t_arg[i] = i;
+        err = pthread_create(&threads[i], NULL, thread_main, (void *) &t_arg[i]);
+        assert(0 == err);
+    }
+
+    for (i = 0; i < T; i++) {
+        int err;
+        err = pthread_join(threads[i], NULL);
+        assert(0 == err);
+    }
+
+    shmem_sync_all();
```

Comment:
Is this sync necessary?  The SHMEM barrier below also appears to be extraneous, but is probably safe programming practice since there isn't a barrier in `main` between initializing `psync` and performing the reduction.  Thus, a change in `thread_main` could introduce a race on `psync`.

Reply:
I think the sync is extra. Added it before I wrote anything in the thread_main and then forgot to remove this one. Removing this. Keeping the barrier though just for safety.